subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/bean/NestedBeanNode.java

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.properties.bean;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import org.codehaus.groovy.runtime.ConvertedClosure;
+import org.gradle.api.GradleException;
+import org.gradle.api.internal.tasks.DefaultTaskInputPropertySpec;
+import org.gradle.api.internal.tasks.PropertySpecFactory;
+import org.gradle.api.internal.tasks.TaskValidationContext;
+import org.gradle.api.internal.tasks.ValidatingValue;
+import org.gradle.api.internal.tasks.ValidationAction;
+import org.gradle.api.internal.tasks.properties.BeanPropertyContext;
+import org.gradle.api.internal.tasks.properties.NodeContext;
+import org.gradle.api.internal.tasks.properties.PropertyMetadata;
+import org.gradle.api.internal.tasks.properties.PropertyMetadataStore;
+import org.gradle.api.internal.tasks.properties.PropertyValue;
+import org.gradle.api.internal.tasks.properties.PropertyValueVisitor;
+import org.gradle.api.internal.tasks.properties.PropertyVisitor;
+import org.gradle.api.internal.tasks.properties.TypeMetadata;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.Optional;
+import org.gradle.internal.Factory;
+import org.gradle.internal.UncheckedException;
+import org.gradle.util.DeferredUtil;
+import org.gradle.util.DeprecationLogger;
+
+import javax.annotation.Nullable;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.List;
+
+import static org.gradle.api.internal.tasks.TaskValidationContext.Severity.ERROR;
+
+class NestedBeanNode extends BaseBeanNode<Object> {
+    public NestedBeanNode(@Nullable String propertyName, Object bean) {
+        super(propertyName, bean);
+    }
+
+    @Override
+    public void visitNode(PropertyVisitor visitor, PropertySpecFactory specFactory, NodeContext<BeanNode> context, PropertyMetadataStore propertyMetadataStore) {
+        if (!isRoot()) {
+            visitImplementation(this, visitor, specFactory);
+        }
+        visitProperties(this, visitor, specFactory, context, propertyMetadataStore);
+    }
+
+    private static void visitImplementation(NestedBeanNode node, PropertyVisitor visitor, PropertySpecFactory specFactory) {
+        // The root bean (Task) implementation is currently tracked separately
```

Comment:
Shouldn't this be on the line with `if (!isRoot())`?

Reply:
The comment is not necessary anymore with the new `RootRuntimeBeanNode`.